File: bkp/ruby/other/procs-lambdas-closures/15-closure-refs.js
var l = console.log.bind(console);

function runIt(callback) {
    callback();
}

var name = 'Yoda';

function printName() {
    l(name);
}

name = 'Luke';

runIt(printName);
// → Luke, just like in ruby.

//
// It seems some languages "lock down" the name value at the point
// the proc/lambda/callback was declared. But in Ruby and JavaScript
// at least, it follows the reference (variable `name` in this case
// and retrieve its value at the point we run the proc/lambda/callback.
//

File: bkp/ruby/other/procs-lambdas-closures/15-closure-refs.rb

def run_it p
    p.call
end

name = 'Yoda'

# You might think, “well, a this point `name` is ‘Yoda’, therefore, later
# when I run this proc it will print ‘Yoda’.”. Nope.
#
# Ruby keeps a reference to the variable, not a copy of its value
# at this point.
#
print_a_name = proc { p name }

name = 'Luke'

run_it print_a_name
    # → Luke, because `name` in the proc is a reference, not a hard copy
# of the value of `name` at the point where the proc was created.
#
